File: backend/billing-service/src/main/java/com/nsbm/group35/healthcare/billing/service/BillingService.java
package com.nsbm.group35.healthcare.billing.service;

import com.nsbm.group35.healthcare.billing.model.Billing;
import com.nsbm.group35.healthcare.billing.repository.BillingRepository;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class BillingService {

    private final BillingRepository billingRepository;

    public BillingService(BillingRepository billingRepository) {
        this.billingRepository = billingRepository;
    }

    public List<Billing> getAllBillings() {
        return billingRepository.findAll();
    }

    public Optional<Billing> getBillingById(String billingId) {
        return billingRepository.findById(billingId);
    }

    public Billing createBilling(Billing billing) {
        return billingRepository.save(billing);
    }

    public Billing updateBilling(String billingId, Billing billing) {
        billing.setBillingId(billingId);
        return billingRepository.save(billing);
    }

    public void deleteBilling(String billingId) {
        billingRepository.deleteById(billingId);
    }

    public List<Billing> getBillingsByPatientId(String patientId) {
        return billingRepository.findByPatientId(patientId);
    }

    public List<Billing> getBillingsByDoctorId(String doctorId) {
        return billingRepository.findByDoctorId(doctorId);
    }

    public List<Billing> getBillingsByAppointmentId(String appointmentId) {
        return billingRepository.findByAppointmentId(appointmentId);
    }

    public List<Billing> getBillingsByPaymentStatus(String paymentStatus) {
        return billingRepository.findByPaymentStatus(paymentStatus);
    }

    public Billing updatePaymentStatus(String billingId, String paymentStatus) {
        Optional<Billing> billing = billingRepository.findById(billingId);
        if (billing.isPresent()) {
            Billing b = billing.get();
            b.setPaymentStatus(paymentStatus);
            return billingRepository.save(b);
        }
        return null;
    }

    public double calculateTotalAmount(double consultationFee, double labTestFee, 
                                       double medicationFee, double otherFee) {
        return consultationFee + labTestFee + medicationFee + otherFee;
    }
}

File: frontend/src/services/billingService.js
import axios from 'axios';

const API_URL = '/api/billings';

// Setup axios instance with dynamic token retrieval
const billingApiClient = axios.create({
    baseURL: API_URL
});

billingApiClient.interceptors.request.use((config) => {
    // Assuming admin or other authed user
    const token = localStorage.getItem('token'); 
    if (token) {
        config.headers.Authorization = `Bearer ${token}`;
    }
    return config;
});

const billingService = {
    // Get all billings
    getAllBillings: async () => {
        const response = await billingApiClient.get('');
        return response.data;
    },

    // Get billing by ID
    getBillingById: async (billingId) => {
        const response = await billingApiClient.get(`/${billingId}`);
        return response.data;
    },

    // Create new billing
    createBilling: async (billingData) => {
        const response = await billingApiClient.post('', billingData);
        return response.data;
    },

    // Update existing billing
    updateBilling: async (billingId, billingData) => {
        const response = await billingApiClient.put(`/${billingId}`, billingData);
        return response.data;
    },

    // Delete a billing
    deleteBilling: async (billingId) => {
        const response = await billingApiClient.delete(`/${billingId}`);
        return response.data;
    },

    // Get billings by patient ID
    getBillingsByPatientId: async (patientId) => {
        const response = await billingApiClient.get(`/patient/${patientId}`);
        return response.data;
    },

    // Get billings by doctor ID
    getBillingsByDoctorId: async (doctorId) => {
        const response = await billingApiClient.get(`/doctor/${doctorId}`);
        return response.data;
    },

    // Get billings by appointment ID
    getBillingsByAppointmentId: async (appointmentId) => {
        const response = await billingApiClient.get(`/appointment/${appointmentId}`);
        return response.data;
    },

    // Get billings by status (PENDING, PAID, etc)
    getBillingsByPaymentStatus: async (paymentStatus) => {
        const response = await billingApiClient.get(`/status/${paymentStatus}`);
        return response.data;
    },

    // Update payment status directly
    updatePaymentStatus: async (billingId, paymentStatus) => {
        // According to backend: @PatchMapping("/{billingId}/status") ... @RequestParam String paymentStatus
        const response = await billingApiClient.patch(`/${billingId}/status`, null, {
            params: { paymentStatus }
        });
        return response.data;
    },

    // Calculate total amount via backend logic
    calculateTotalAmount: async (consultationFee, labTestFee, medicationFee, otherFee) => {
        const response = await billingApiClient.post('/calculate', null, {
            params: {
                consultationFee,
                labTestFee,
                medicationFee,
                otherFee
            }
        });
        return response.data;
    }
};

export default billingService;
